src/objective/rank_xendcg_objective.hpp

```diff
@@ -0,0 +1,162 @@
+#ifndef LIGHTGBM_OBJECTIVE_RANK_XENDCG_OBJECTIVE_HPP_
+#define LIGHTGBM_OBJECTIVE_RANK_XENDCG_OBJECTIVE_HPP_
+
+#include <LightGBM/objective_function.h>
+
+#include <limits>
+#include <string>
+#include <algorithm>
+#include <cmath>
+#include <cstdio>
+#include <cstring>
+#include <math.h>
+#include <random>
+#include <vector>
+#include <random>
+
+namespace LightGBM {
+/*!
+* \brief Implementation of the learning-to-rank objective function, XE_NDCG [arxiv.org/abs/1911.09798].
+*/
+class RankXENDCG: public ObjectiveFunction {
+ public:
+  explicit RankXENDCG(const Config& config) :
+      generator_((std::random_device())()), distribution_(0.0, 1.0) {
+    if (config.seed != 0) {
+      generator_.seed(config.seed);
+    }
+  }
+
+  explicit RankXENDCG(const std::vector<std::string>&) :
+      generator_((std::random_device())()), distribution_(0.0, 1.0) {}
+
+  ~RankXENDCG() {
+  }
+  void Init(const Metadata& metadata, data_size_t) override {
+    // get label
+    label_ = metadata.label();
+    // get boundries
+    query_boundaries_ = metadata.query_boundaries();
+    if (query_boundaries_ == nullptr) {
+      Log::Fatal("RankXENDCG tasks require query information");
+    }
+    num_queries_ = metadata.num_queries();
+  }
+
+  void GetGradients(const double* score, score_t* gradients,
+                    score_t* hessians) const override {
+    #pragma omp parallel for schedule(guided)
+    for (data_size_t i = 0; i < num_queries_; ++i) {
+      GetGradientsForOneQuery(score, gradients, hessians, i);
+    }
+  }
+
+  inline void GetGradientsForOneQuery(
+      const double* score,
+      score_t* lambdas, score_t* hessians, data_size_t query_id) const {
+    // get doc boundary for current query
+    const data_size_t start = query_boundaries_[query_id];
+    const data_size_t cnt =
+      query_boundaries_[query_id + 1] - query_boundaries_[query_id];
+    // add pointers with offset
+    const label_t* label = label_ + start;
+    score += start;
+    lambdas += start;
+    hessians += start;
+
+    // Turn scores into a probability distribution using Softmax.
```

Comment:
I remember LightGBM already has `Softmax` implementation:
https://github.com/microsoft/LightGBM/blob/d93984345445e09cea083032e85c17163edcb789/include/LightGBM/utils/common.h#L598-L631

Reply:
Thanks! I've updated the code to use Common::Softmax().